tests/react/async.test.tsx

```diff
@@ -1143,3 +1143,142 @@ it('multiple derived atoms with dependency chaining and async write (#813)', asy
     getByText('bName: beta')
   })
 })
+
+it('[render] resolves dependencies reliably after a delay', async () => {
+  expect.assertions(2)
+  const countAtom = atom(0)
+  const resultAtom = atom<number | null>(null)
+
+  const resolve: (() => void)[] = []
+  const asyncAtom = atom(async (get) => {
+    const count = get(countAtom)
+    await new Promise<void>((r: { (): void; count?: number }) => {
+      r.count = count
+      resolve.push(r)
+    })
+    console.log(`resolved (${count})`)
+    return count
+  })
+
+  const derivedAtom = atom(
+    async (get, { setSelf }) => {
+      const count = get(countAtom)
+      await Promise.resolve()
+      console.log(`derived (${count})`)
+      const resultCount = await get(asyncAtom)
+      console.log(`derived (${count})`, resultCount)
+      setSelf(resultAtom, resultCount)
```

Comment:
Please do not use `setSelf` if we can reproduce the issue without it.

(Otherwise, we have a different file `async2.test.tsx` for v2 specific apis.)

Reply:
This is weird. When I comment out the setSelf line, the test passes.
```ts
const derivedAtom = atom(
    async (get, { setSelf }) => {
      const count = get(countAtom)
      await Promise.resolve()
      console.log(`derived (${count})`)
      const resultCount = await get(asyncAtom)
      console.log(`derived (${count})`, resultCount)
      result = resultCount
      setSelf(resultAtom, resultCount) // <--- comment out this line and the test fails ¯\_(ツ)_/¯
    },
    (_get, set, ...args: Parameters<Setter>) => set(...args)
  )
...
expect(result).toBe(4) // 3
```

Reply:
fails
```ts
if (resultCount === 2) setSelf(resultAtom, resultCount)
```

```
derived (0)
resolved (0)
derived (1)
derived (2)
resolved (1)
resolved (2)
derived (2) 2
derived (1) 2
derived (0) 2
derived (3)
derived (4)
resolved (3)
resolved (4)
derived (3) 3 <-- should be 4
derived (4) 3
```

Reply:
passes
```ts
if (resultCount === 4) setSelf(resultAtom, resultCount)
```

```
derived (0)
resolved (0)
derived (1)
derived (2)
resolved (1)
resolved (2)
derived (2) 2
derived (1) 2
derived (0) 2
derived (3)
derived (4)
resolved (3)
resolved (4)
derived (4) 4 <-- passes
derived (3) 4
```

Reply:
Since `setSelf`, itself does not introduce any micro or macro task delay. This difference between fails / passes is the result of a side-effect of setSelf and not actually caused by a race condition.

Although there could still be an async side-effect of setSelf that is racing.